client/src/controllers/SlugController.ts

```diff
@@ -10,13 +10,43 @@ import { cleanForSlug } from '../utils/text';
 export class SlugController extends Controller<HTMLInputElement> {
   static values = {
     allowUnicode: { default: false, type: Boolean },
+    applyUpdates: { default: false, type: Boolean },
   };
 
   declare allowUnicodeValue: boolean;
+  declare applyUpdatesValue: boolean;
 
-  slugify() {
-    this.element.value = cleanForSlug(this.element.value.trim(), false, {
+  compare(event) {
+    const target = this.element.value;
+
+    const compareMethod = () =>
+      event.params.compareAs === 'slugify'
+        ? this.slugify()
+        : cleanForSlug(event.detail.value.trim(), true);
+    this.applyUpdatesValue = compareMethod() === target;
+
+    if (this.applyUpdatesValue) {
+      event.preventDefault();
+    }
+  }
+
+  slugify(applyUpdate = false) {
+    const slugifiedValue = cleanForSlug(this.element.value.trim(), false, {
```

Comment:
It looks like you are calling `compareMethod` here twice, this is probably why it's not working as expected.

Here is an updated controller that takes this code so far and updates it to have a consistent urlify/slugify method.

* `compare` - only attempt to check if the controlled element's value is not blank, then compare against the transformed supplied value with the controlled element's value. Note there is a new second param on both `urlify` and `slugify` that allows the transform to happen but NOT update the field's value. This is instead of trying to track `applyUpdatesValue` as state.
* `urlify`/`slugify` - both of these should allow for the providing of the value to use from the `event.detail.value` with a fallback to the controlled element's value. Then the second param determines if the new transformed value actually gets put onto the field.

```ts
export class SlugController extends Controller<HTMLInputElement> {
  static values = {
    allowUnicode: { default: false, type: Boolean },
  };

  declare allowUnicodeValue: boolean;

  /**
   * Allow for a comparison value to be provided, if does not compare to the
   * current value (once transformed), then the event's default will
   * be prevented.
   */
  compare(
    event: CustomEvent<{ value: string }> & { params?: { compareAs?: string } },
  ) {
    // do not attempt to compare if the current field is empty
    if (!this.element.value) {
      return true;
    }

    const {
      detail: { value },
      params: { compareAs = 'slugify' } = {},
    } = event;

    const compareValue = this[compareAs]({ detail: { value } }, true);
    const currentValue = this.element.value;

    const valuesAreSame = compareValue.trim() === currentValue.trim();

    if (valuesAreSame) {
      event?.preventDefault();
    }

    return valuesAreSame;
  }

  /**
   * Basic slugify of a string, updates the controlled element's value
   * or can be used to simply return the transformed value.
   * If a custom event with detail.value is provided, that value will be used
   * instead of the field's value.
   */
  slugify(event: CustomEvent<{ value: string }>, ignoreUpdate = false) {
    const unicodeSlugsEnabled = this.allowUnicodeValue;
    const { value = this.element.value } = event?.detail || {};
    const newValue = cleanForSlug(value.trim(), false, { unicodeSlugsEnabled });

    if (!ignoreUpdate) {
      this.element.value = newValue;
    }

    return newValue;
  }

  /**
   * Advanced slugify of a string, updates the controlled element's value
   * or can be used to simply return the transformed value.
   * If a custom event with detail.value is provided, that value will be used
   * instead of the field's value.
   */
  urlify(event: CustomEvent<{ value: string }>, ignoreUpdate = false) {
    const unicodeSlugsEnabled = this.allowUnicodeValue;
    const { value = this.element.value } = event?.detail || {};
    const newValue = cleanForSlug(value.trim(), true, { unicodeSlugsEnabled });

    if (!ignoreUpdate) {
      this.element.value = newValue;
    }

    return newValue;
  }
}
```

Reply:
At what point would the `params` method change to `urlify` since the default is `slugify` method and we aren't setting the params in the data-attributes?

Reply:
at no point is this needed to preserve current state, so we *could* hard-code it, but allowing it to be overridden by an event param makes it easy to test and also adds support for future usage of slugify only behaviour for comparison.

Reply:
Hmm
Makes ense